Cython/Compiler/PyrexTypes.py

```diff
@@ -1345,28 +1349,43 @@ def generate_xdecref_clear(self, code, cname, clear_before_decref=False, nanny=T
                          clear=True, clear_before_decref=clear_before_decref)
 
     def generate_gotref(self, code, cname):
+        if code.funcstate:
+            code.funcstate.needs_refnanny = True
         code.putln("__Pyx_GOTREF(%s);" % self.as_pyobject(cname))
 
     def generate_xgotref(self, code, cname):
+        if code.funcstate:
+            code.funcstate.needs_refnanny = True
         code.putln("__Pyx_XGOTREF(%s);" % self.as_pyobject(cname))
 
     def generate_giveref(self, code, cname):
+        if code.funcstate:
+            code.funcstate.needs_refnanny = True
         code.putln("__Pyx_GIVEREF(%s);" % self.as_pyobject(cname))
 
     def generate_xgiveref(self, code, cname):
+        if code.funcstate:
+            code.funcstate.needs_refnanny = True
         code.putln("__Pyx_XGIVEREF(%s);" % self.as_pyobject(cname))
 
     def generate_decref_set(self, code, cname, rhs_cname):
+        if code.funcstate:
+            code.funcstate.needs_refnanny = True
         code.putln("__Pyx_DECREF_SET(%s, %s);" % (cname, rhs_cname))
 
     def generate_xdecref_set(self, code, cname, rhs_cname):
+        if code.funcstate:
+            code.funcstate.needs_refnanny = True
         code.putln("__Pyx_XDECREF_SET(%s, %s);" % (cname, rhs_cname))
 
     def _generate_decref(self, code, cname, nanny, null_check=False,
                     clear=False, clear_before_decref=False):
         prefix = '__Pyx' if nanny else 'Py'
         X = 'X' if null_check else ''
 
+        if code.funcstate:
+            code.funcstate.needs_refnanny = code.funcstate.needs_refnanny or nanny
```

Comment:
```suggestion
        if code.funcstate and nanny:
            code.funcstate.needs_refnanny = True
```

Reply:
It looks to be in the module deallocation functions. `nanny` is set to false though, so testing it first is enough to not need the funcstate check I think.